Update library to Python 3.5+ and asyncio

bachya opened the pull request:
Hi there! Following up on our conversation [here](https://github.com/home-assistant/home-assistant/pull/17535), this PR updates the library to be `asyncio`-friendly. It also updates the architecture to be more modular and friendly for future additions.

In addition to the library changes, this PR adds a few nice tooling changes that make future contributions easier:

- A Pipfile
- A Makefile
- Various tools to check linting, typing, etc.
- A more standard `setup.py`

Lastly, this bumps the version to `1.0.0`. I chose that number since this is a very breaking update and usually, the first digit gets bumped in that case. Totally open to your wishes, though.

My suggestion as to how to proceed:

1. Let's work through this PR and get it merged.
2. Let's open a new Home Assistant PR that focuses simply on transitioning the existing functionality to this new version (I have one ready to go).
3. You rebase https://github.com/home-assistant/home-assistant/pull/17535 on top of step 2, fix any issues that arise, and proceed with getting that PR merged.

Thoughts?

bachya committed:
Initial code in place

bachya committed:
Final code in place

bachya committed:
Updated README.md

bachya committed:
Remove unnecessary file

bachya committed:
Updated example

ehendrix23 commented:
@arraylabs, @bachya .

Would it make sense to use requests.session instead of requests.get and requests.post? That would allow the TCP connection to stay open and thus no need to keep on recreating it.
I just submitted a PR for august (
https://github.com/snjoetw/py-august/pull/11) and it was fairly easy to do. Looking at MyQ I would say it should be fairly simple here as well.

The difference for August from my home was 294ms with request.get compared to 80ms for request.session. Thoughts?

bachya commented:
@ehendrix23 That's definitely the correct principle. Note that this PR does away with `requests` altogether: since it attempts to make the library `asyncio`-compatible, it replaces `requests` with `aiohttp`. You can see [here](https://github.com/arraylabs/pymyq/blob/04e60a20c34687667b8fea0ab3e696d39f6ec1cd/pymyq/api.py#L49) that it uses a `ClientSession`, which accomplishes what you're talking about.

ehendrix23 commented:
@bachya Well, that teaches me to just have a look at existing code and the description of this PR. :-)

ehendrix23 committed:
Added retries for open & close

Added retries for open and close as I've seen this fail as well.
Changed so that update, open, and close return True if successful or False if unsuccessful.

bachya committed:
Merge pull request #1 from ehendrix23/dev-update

Added retry for open/close and True/False return for update, close, open.

vajonam commented:
can I help with this? would like to get this back into home assistant as working component.

bachya commented:
@vajonam: Waiting on @arraylabs to approve.

arraylabs reviewed (approved):
Wow, that appears to have been an undertaking! Will take me a bit to wrap my head around all of it. Thanks for the hard work and I apologize for the delay in getting back to it as I've been traveling again.

bachya commented:
No problem, @arraylabs! Are you good if I start on number 2 in my checklist above?

arraylabs commented:
@bachya yes please do!

bachya commented:
I’ll need you to publish a new version on PyPi. Let me know when that’s in place. 👍🏻

arraylabs commented:
@bachya it should be there now, fyi adding the publish/upload to the setup.py just makes things too easy. I should have thought about that about 15 versions ago...

ehendrix23 commented:
@arraylabs if you would have done it 15 versions ago then you would not be appreciating how easy that makes it now. :-)

@bachya, let me know when you have something for testing. As long as my garage doors won't start opening and closing on random moments I'm good. :-)

bachya commented:
@ehendrix23 FYI, I've submitted a PR to HASS: https://github.com/home-assistant/home-assistant/pull/18489 – you can use that for testing.

ehendrix23 commented:
@bachya,

OK, just put it in my HASS. Working so far. :-)

Noticed though that there are no debug logger messages. I might see to add some in both pymyq and the HASS component once we've got this PR in. 
Those can help when something isn't working as it should. :-)

vajonam commented:
I am using this as well, works fine this far. no issues to report
